### PR TITLE
add json_patch, json_patch_recipe and json_diff filters

### PR DESCRIPTION
**Pull request by numo68**
##### SUMMARY

The Ansible collections lack a good way to modify JSON data. This PR leverages the [jsonpatch](https://pypi.org/project/jsonpatch/) package and adds three filters implementing [RFC 6902](https://datatracker.ietf.org/doc/html/rfc6902) operations on an object.

##### ISSUE TYPE

- New Module/Plugin Pull Request

##### COMPONENT NAME

JavaScript Object Notation (JSON) Patch Filter

##### ADDITIONAL INFORMATION

There is a previous work done in the (https://github.com/ParticleDecay/ansible-jsonpatch) repo, but I think that the operations themselves should be available on arbitrary data, not just on files; a file modification can be added as a role if necessary.

I am a moderate Ansible user, but the first time contributor to Ansible collections, so if I forgot something, please kindly point out the deficiencies and I'll do my best to fix it.

Thanks for a great tool and great modules!

### CODE REVIEW

File: plugins/filter/json_diff.yml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+DOCUMENTATION:
+  name: json_diff
+  short_description: Create a JSON patch by comparing two JSON files
+  description:
+    - This filter compares the input with the argument and computes a list of operations
+      that can be consumed by the P(community.general.json_patch_recipe) to change the input
+      to the argument.
+  requirements:
+    - jsonpatch (on the controller)
```
Comment (felixfontein) [inline plugins/filter/json_diff.yml:14]: The comment on controller isn't needed since it is added automatically, see https://docs.ansible.com/ansible/devel/collections/community/general/jc_filter.html#requirements: "The below requirements are needed on the local controller node that executes this filter."
```suggestion
    - jsonpatch
```

File: plugins/filter/json_diff.yml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        type: raw
+
+EXAMPLES: |
+  - name: Compute a difference
+    ansible.builtin.debug:
+      msg: "{{ input | community.general.json_diff(target) }}"
+    vars:
+      input: {"foo": 1, "bar":{"baz": 2}, baw: [1, 2, 3], "hello": "day"}
+      target: {"foo": 1, "bar": {"baz": 2}, baw: [1, 3], "baq": {"baz": 2}, "hello": "night"}
+    # => [{"op": "add", "path": "/baq", "value": {"baz": 2}}, {"op": "remove", "path": "/baw/1"}, {"op": "replace", "path": "/hello", "value": "night"}]
```
Comment (felixfontein) [inline plugins/filter/json_diff.yml:63]: To increase readability (and avoid horizontal scrolling), how about splitting this up as follows:
```suggestion
    # => [
    #   {"op": "add", "path": "/baq", "value": {"baz": 2}},
    #   {"op": "remove", "path": "/baw/1"},
    #   {"op": "replace", "path": "/hello", "value": "night"}
    # ]
```

File: plugins/filter/json_patch.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+if TYPE_CHECKING:
+    from typing import Callable
+
+try:
+    import jsonpatch
+
+    HAS_LIB = True
+except ImportError:
+    HAS_LIB = False
```
Comment (felixfontein) [inline plugins/filter/json_patch.py:22]:
```suggestion
except ImportError as exc:
    HAS_LIB = False
    JSONPATCH_IMPORT_ERROR = exc
else:
    HAS_LIB = True
    JSONPATCH_IMPORT_ERROR = None
```

File: plugins/filter/json_patch.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+class FilterModule:
+    """Filter plugin."""
+
+    def json_patch(
+        self, inp: object, op: str, path: str, value: object = None, **kwargs: dict
+    ) -> object:
+
+        if not HAS_LIB:
+            raise AnsibleError(
```
Comment (felixfontein) [inline plugins/filter/json_patch.py:33]: Better use `AnsibleFilterError` for errors reported by filters:
```suggestion
            raise AnsibleFilterError(
```

File: plugins/filter/json_patch.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                'You need to install "jsonpatch" package prior to running "json_patch" filter'
+            )
```
Comment (felixfontein) [inline plugins/filter/json_patch.py:35]:
```suggestion
            ) from JSONPATCH_IMPORT_ERROR
```
see https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/import.html#in-plugins

File: tests/unit/plugins/filter/test_json_diff.py
Comment (felixfontein) [inline tests/unit/plugins/filter/test_json_diff.py:1]: The file must be called `test_json_patch.py` since it tests functionality in `json_patch.py`.

**Review by felixfontein — COMMENTED** [review]
Thanks for your contriubtion! I've added some first comments.

Can you also add integration tests (`tests/integration/targets/filter_json_patch/`; see for example `tests/integration/targets/filter_jc/` for how to install the required library before running the tests)?

**Comment by felixfontein** [comment]
Regarding BOTMETA, see https://github.com/ansible-collections/community.general/blob/main/CONTRIBUTING.md#creating-new-modules-or-plugins.

File: tests/unit/plugins/filter/test_json_diff.py
Comment (numo68) [reply tests/unit/plugins/filter/test_json_diff.py:1]: Hmm, there are three filters, is it not possible to split the tests into three files for clarity?

Thank you for your comments and suggestions, will implement it soon.

File: plugins/filter/json_diff.yml [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+# Copyright (c) Stanislav Meduna (@numo68)
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+DOCUMENTATION:
+  name: json_diff
+  short_description: Create a JSON patch by comparing two JSON files
+  description:
+    - This filter compares the input with the argument and computes a list of operations
+      that can be consumed by the P(community.general.json_patch_recipe) to change the input
```
Comment (felixfontein) [inline plugins/filter/json_diff.yml:11]:
```suggestion
      that can be consumed by the P(community.general.json_patch_recipe#filter) filter to change the input
```

File: tests/unit/plugins/filter/test_json_diff.py
Comment (felixfontein) [reply tests/unit/plugins/filter/test_json_diff.py:1]: Then you need to put them in a subdirectory `json_patch`, I think.

File: plugins/filter/json_patch.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+        args = {"op": op, "path": path}
```
Comment (felixfontein) [inline plugins/filter/json_patch.py:37]: Please note that you have to validate the inputs, i.e. make sure the strings are actually strings, `op` is one of the required strings, and so on.

File: plugins/filter/json_diff.yml [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+      to the argument.
+  requirements:
+    - jsonpatch (on the controller)
+  version_added: 10.3.0
+  author:
+    - Stanislav Meduna (@numo68)
```
Comment (felixfontein) [inline plugins/filter/json_diff.yml:17]:
```suggestion
    - Stanislav Meduna (@numo68)
  positional: ...
```
You need to list the positional arguments in order here. Otherwise the users have no clue in which order the options must be supplied in case they're not keyword options.

File: plugins/filter/json_patch.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    HAS_LIB = True
+except ImportError:
+    HAS_LIB = False
+
+
+class FilterModule:
+    """Filter plugin."""
+
+    def json_patch(
+        self, inp: object, op: str, path: str, value: object = None, **kwargs: dict
```
Comment (felixfontein) [inline plugins/filter/json_patch.py:29]: `object` is a bad choice for input. You do not want arbitrary Python objects here, but you want strings or mappings. So better start with `typing.Any` and check what you got.
```suggestion
        self, inp: typing.Any, op: str, path: str, value: object = None, **kwargs: dict
```
Also why use `**kwargs: dict` instead of `/, from: str | None = None`?

Comment (numo68) [reply plugins/filter/json_patch.py:29]: > Also why use `**kwargs: dict` instead of `/, from: str | None = None`?

`from` is a reserved word

Comment (felixfontein) [reply plugins/filter/json_patch.py:29]: In that case you have to make sure that `kwargs` doesn't contain anything else than `from` (or what else you might extract from it).

File: plugins/filter/json_patch.py (hunk 7 above)
Comment (numo68) [reply plugins/filter/json_patch.py:37]: > Please note that you have to validate the inputs, i.e. make sure the strings are actually strings, `op` is one of the required strings, and so on.

How deep should that validation go? For example the `json_patch_recipe` accepts an array of such operations, do I have to check all? The underlying `jsonpatch` library will check and bail out if unacceptable, so I don't think the second level is necessary, but I don't know what the common way of handling in ansible is.

Comment (felixfontein) [reply plugins/filter/json_patch.py:37]: I think a general guideline is to at least validate what you are describing in the plugin's documentation. Whether you want to validate more depends on how nice the error messages should be (which depends on how user-friendly the errors returned from the library are - if they simply say "bad input" doing more validation is a good idea, or if they use other parameter names than the ones supplied to the plugin).

**Comment by numo68** [comment]
- comments resolved
- input validation added
- integration test added

File: plugins/filter/json_patch.yml [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+DOCUMENTATION:
+  name: json_patch
+  short_description: Apply a JSON-Patch (RFC 6902) operation to an object
+  description:
+    - This filter applies a single JSON patch operation and returns a modified object.
+    - If the operation is a test, the filter returns an ummodified object if the test
+      succeeded and a None value otherwise.
```
Comment (felixfontein) [inline plugins/filter/json_patch.yml:12]: On Ansible's side the Python `None` is usually denoted by `none` or `null`.
```suggestion
      succeeded and a V(none) value otherwise.
```

File: plugins/filter/json_patch.yml [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+      description: Operation to perform (see RFC 6902).
+      type: str
+      choices: [add, copy, move, remove, replace, test]
+      required: true
+    path:
+      description: JSON Pointer path to the target location (see RFC 6901).
+      type: str
+      required: true
+    value:
+      description: Value to use in the operation. Ignored for copy, move and remove.
```
Comment (felixfontein) [inline plugins/filter/json_patch.yml:34]:
```suggestion
      description: Value to use in the operation. Ignored for O(op=copy), O(op=move), and O(op=remove).
```

File: plugins/filter/json_patch.yml [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+      description: A list or a dictionary representing a JSON object, or a string containing a JSON object.
+      type: raw
+      required: true
+    op:
+      description: Operation to perform (see RFC 6902).
+      type: str
+      choices: [add, copy, move, remove, replace, test]
+      required: true
+    path:
+      description: JSON Pointer path to the target location (see RFC 6901).
```
Comment (felixfontein) [inline plugins/filter/json_patch.yml:30]:
```suggestion
      description: JSON Pointer path to the target location (see L(RFC 6901, https://datatracker.ietf.org/doc/html/rfc6901)).
```

File: plugins/filter/json_patch.yml [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+      type: str
+      required: true
+    value:
+      description: Value to use in the operation. Ignored for copy, move and remove.
+      type: raw
+    from:
+      description: The source location for the copy and move operation. Mandatory
+        for copy and move operations, ignored otherwise. Keyword argument only.
```
Comment (felixfontein) [inline plugins/filter/json_patch.yml:38]:
```suggestion
        for O(op=copy) and O(op=move), ignored otherwise. Keyword argument only.
```

File: plugins/filter/json_patch.yml [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+    - name: RFC 6901
+      description: JavaScript Object Notation (JSON) Pointer
+      link: https://datatracker.ietf.org/doc/html/rfc6901
+    - name: jsonpatch Python Package
+      description: A Python library for applying JSON patches
+      link: https://pypi.org/project/jsonpatch/
+
+RETURN:
+  _value:
+    description: A modified object or None if the operation is a test and the test failed.
```
Comment (felixfontein) [inline plugins/filter/json_patch.yml:53]:
```suggestion
    description: A modified object or V(none) if the operation is a test and the test failed.
```

File: plugins/filter/json_patch_recipe.yml [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+DOCUMENTATION:
+  name: json_patch_recipe
+  short_description: Apply JSON-Patch (RFC 6902) operations to an object
+  description:
+    - This filter sequentially applies JSON patch operations and returns a modified object.
+    - If there is a test operation in the list, the filter continues if the test
+      succeeded and returns a None value otherwise.
```
Comment (felixfontein) [inline plugins/filter/json_patch_recipe.yml:12]:
```suggestion
      succeeded and returns a V(none) value otherwise.
```

File: plugins/filter/json_patch_recipe.yml [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+          description: Operation to perform (see RFC 6902).
+          type: str
+          choices: [add, copy, move, remove, replace, test]
+          required: true
+        path:
+          description: JSON Pointer path to the target location (see RFC 6901).
+          type: str
+          required: true
+        value:
+          description: Value to use in the operation. Ignored for copy, move and remove.
```
Comment (felixfontein) [inline plugins/filter/json_patch_recipe.yml:40]:
```suggestion
          description: Value to use in the operation. Ignored for O(operations[].op=copy), O(operations[].op=move), and O(operations[].op=remove).
```

File: plugins/filter/json_patch_recipe.yml [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+          type: raw
+        from:
+          description: The source location for the copy and move operation. Mandatory
+            for copy and move operations, ignored otherwise.
```
Comment (felixfontein) [inline plugins/filter/json_patch_recipe.yml:44]:
```suggestion
            for O(operations[].op=copy) and O(operations[].op=move), ignored otherwise.
```

File: plugins/filter/json_patch_recipe.yml [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+    - name: RFC 6901
+      description: JavaScript Object Notation (JSON) Pointer
+      link: https://datatracker.ietf.org/doc/html/rfc6901
+    - name: jsonpatch Python Package
+      description: A Python library for applying JSON patches
+      link: https://pypi.org/project/jsonpatch/
+
+RETURN:
+  _value:
+    description: A modified object or None if the operation is a test and the test failed.
```
Comment (felixfontein) [inline plugins/filter/json_patch_recipe.yml:59]:
```suggestion
    description: A modified object or V(none) if the operation is a test and the test failed.
```

File: plugins/filter/json_patch.py [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+class FilterModule:
+    """Filter plugin."""
+
+    def check_json_object(
+        self, filter_name: str, object_name: str, inp: Union[str, list, dict]
+    ):
+        if not isinstance(inp, (str, list, dict)):
+            raise AnsibleFilterError(
+                f"{filter_name}: {object_name} is not dictionary, list or string"
+            )
```
Comment (felixfontein) [inline plugins/filter/json_patch.py:41]: I think it would be a good idea if `check_json_object` would already call `loads()` if the input is a string, and handle exceptions that `loads()` throws. That would simplify the callers of this method considerably.

File: plugins/filter/json_patch.py [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+        value: Any = None,
+        **kwargs: dict,
+    ) -> Any:
+
+        if not HAS_LIB:
+            raise AnsibleFilterError(
+                "You need to install 'jsonpatch' package prior to running 'json_patch' filter"
+            ) from JSONPATCH_IMPORT_ERROR
+
+        self.check_json_object("json_patch", "input", inp)
```
Comment (felixfontein) [inline plugins/filter/json_patch.py:80]: This could be...
```suggestion
        inp = self.check_json_object("json_patch", "input", inp)
```

File: plugins/filter/json_patch.py [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+        if op in OPERATIONS_NEEDING_FROM and from_arg is not None:
+            args["from"] = from_arg
+
+        self.check_patch_arguments("json_patch", args)
+
+        result = None
+
+        try:
+            result = jsonpatch.apply_patch(
+                inp if isinstance(inp, (dict, list)) else loads(inp),
```
Comment (felixfontein) [inline plugins/filter/json_patch.py:101]: ... and this:
```suggestion
                inp,
```

File: plugins/filter/json_patch.py [hunk 22]
```diff
@@ -0,0 +1,176 @@
...
+OPERATIONS_NEEDING_VALUE = ["add", "replace", "test"]
+
+
+class FilterModule:
+    """Filter plugin."""
+
+    def check_json_object(
+        self, filter_name: str, object_name: str, inp: Union[str, list, dict]
+    ):
+        if not isinstance(inp, (str, list, dict)):
```
Comment (felixfontein) [inline plugins/filter/json_patch.py:38]: I would probably also accept `binary` here. `loads` also handles binary strings, and if the filter is chained with something that outputs a binary string (like the `ansible.builtin.b64decode` filter), this would make life a lot simpler for the user.

File: plugins/filter/json_patch.py [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+            ) from JSONPATCH_IMPORT_ERROR
+
+        self.check_json_object("json_patch", "input", inp)
+
+        args = {"op": op, "path": path}
+        from_arg = kwargs.pop("from", None)
+
+        if kwargs:
+            raise AnsibleFilterError(
+                f"json_patch: unexpected keywords arguments: {', '.join(kwargs.keys())}"
```
Comment (felixfontein) [inline plugins/filter/json_patch.py:87]: Sorting the keys makes the output more predictable (could be useful for tests):
```suggestion
                f"json_patch: unexpected keywords arguments: {', '.join(sorted(kwargs))}"
```

File: plugins/filter/json_patch.py [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+
+        try:
+            result = jsonpatch.apply_patch(
+                inp if isinstance(inp, (dict, list)) else loads(inp),
+                [args],
+            )
+        except jsonpatch.JsonPatchTestFailed:
+            pass
+        except Exception as e:
+            raise AnsibleFilterError(f"json_patch: patch failed: {to_native(e)}") from e
```
Comment (felixfontein) [inline plugins/filter/json_patch.py:107]: `to_native` is no longer necessary for plugins (since they no longer support Python 2.7). You can simply use `str()` for this use-case, or (even better) simply leave it away since f-strings automatically call `str()` on non-strings:
```suggestion
            raise AnsibleFilterError(f"json_patch: patch failed: {e}") from e
```

File: plugins/filter/json_patch.py [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+OPERATIONS_AVAILABLE = ["add", "copy", "move", "remove", "replace", "test"]
+OPERATIONS_NEEDING_FROM = ["copy", "move"]
+OPERATIONS_NEEDING_VALUE = ["add", "replace", "test"]
+
+
+class FilterModule:
+    """Filter plugin."""
+
+    def check_json_object(
+        self, filter_name: str, object_name: str, inp: Union[str, list, dict]
```
Comment (felixfontein) [inline plugins/filter/json_patch.py:36]: Why not use `Any` here? After all you're verifying the type in the function.
```suggestion
        self, filter_name: str, object_name: str, inp: Any
```

File: plugins/filter/json_patch.yml [hunk 26]
```diff
@@ -0,0 +1,140 @@
...
+  positional: op, path, value
+  options:
+    _input:
+      description: A list or a dictionary representing a JSON object, or a string containing a JSON object.
+      type: raw
+      required: true
+    op:
+      description: Operation to perform (see RFC 6902).
+      type: str
+      choices: [add, copy, move, remove, replace, test]
```
Comment (felixfontein) [inline plugins/filter/json_patch.yml:27]: I think it would be useful if this filter and `json_patch_recipe` have an option which allows to turn failed `test`s into proper errors (`AnsibleFilterError` exception). Otherwise you'll always have to write extra code in the playbook to make sure that the result isn't `none` before using the result of these filters.

File: tests/integration/targets/filter_json_patch/tasks/main.yml [hunk 27]
```diff
@@ -0,0 +1,139 @@
...
+####################################################################
+# WARNING: These are designed specifically for Ansible tests       #
+# and should not be used as examples of how to write Ansible roles #
+####################################################################
+
+# Copyright (c) Ansible Project
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+- name: test json_patch
```
Comment (felixfontein) [inline tests/integration/targets/filter_json_patch/tasks/main.yml:11]: Task names should start with an upper-case letter:
```suggestion
- name: Test json_patch
```

File: tests/integration/targets/filter_json_patch/tasks/main.yml [hunk 28]
```diff
@@ -0,0 +1,139 @@
...
+      - op: 'copy'
+        from: '/baz'
+        path: '/bay'
+      - op: 'replace'
+        path: '/baz'
+        value: [10, 20, 30]
+
+- name: test json_diff
+  assert:
+    that: # The order in the result array is not stable, check per key
```
Comment (felixfontein) [inline tests/integration/targets/filter_json_patch/tasks/main.yml:124]: How about sorting the result by key `path`? Then you can simply compare the output to a (fixed-order) list.

File: tests/unit/plugins/filter/test_json_patch.py [hunk 29]
```diff
@@ -0,0 +1,318 @@
...
+
+    # json_patch
+
+    def test_patch_add_to_empty(self):
+        result = json.dumps(
+            self.json_patch({}, "add", "/a", 1),
+            sort_keys=True,
+            separators=(",", ":"),
+        )
+        self.assertEqual(result, '{"a":1}')
```
Comment (felixfontein) [inline tests/unit/plugins/filter/test_json_patch.py:31]: Why dump the result and comparing JSON strings, instead of comparing objects:
```suggestion
        self.assertEqual(result, {"a": 1})
```
?

File: plugins/filter/json_patch.py (hunk 22 above)
Comment (numo68) [reply plugins/filter/json_patch.py:38]: Hmm, according to the documentation and also a quick test the `b64decode` outputs a string, not `bytes` or `bytearray`. If there is any way to pass a binary input to the filter, I am not aware of it - is there?

File: plugins/filter/json_patch.yml (hunk 26 above)
Comment (bcoca) [reply plugins/filter/json_patch.yml:27]: There is, but only from inside another Jinja plugin and while using the 'native' engine. From Ansible itself you cannot pass byte strings, only unicode strings.

File: plugins/filter/json_patch.py (hunk 22 above)
Comment (bcoca) [reply plugins/filter/json_patch.py:38]: There is, but only from inside another Jinja plugin and while using the 'native' engine. From Ansible itself you cannot pass byte strings, only unicode strings.

Comment (numo68) [reply plugins/filter/json_patch.py:38]: okay in that case I'll implement accepting bytes and bytearray in the python and test it in the unit test, but no further comment on it in docs nor errors returned.

**Comment by numo68** [comment]
Many thanks for the valuabe input. I have addressed the - hopefully all - comments. The filter now accepts binary data (even if not possible to pass it from the play) and there is a `fail_test` argument to specify whether a failed test returns `none` or fails. The default is still `none` as it will be IMO more used to take decisions than to check validity.

Thanks

**Review by russoz — COMMENTED** [review]
Hi @numo68 Thanks for contributing!

Got a couple of comments.

File: plugins/filter/json_patch.yml [hunk 30]
```diff
@@ -0,0 +1,145 @@
...
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+DOCUMENTATION:
+  name: json_patch
+  short_description: Apply a JSON-Patch (RFC 6902) operation to an object
+  description:
+    - This filter applies a single JSON patch operation and returns a modified object.
+    - If the operation is a test, the filter returns an ummodified object if the test
+      succeeded and a V(none) value otherwise.
```
[the remark this answers is not in the document]
Comment (felixfontein) [reply plugins/filter/json_patch.yml:12]: Actually `null` is JSON, while `none` is Jinja2. Since this is a Jinja2 filter, I think `none` is more appropriate.

File: plugins/filter/json_patch.py [hunk 31]
```diff
@@ -0,0 +1,195 @@
...
+                "You need to install 'jsonpatch' package prior to running 'json_patch_recipe' filter"
+            ) from JSONPATCH_IMPORT_ERROR
+
+        if not isinstance(operations, list):
+            raise AnsibleFilterError(
+                "json_patch_recipe: 'operations' needs to be a list"
+            )
+
+        if not isinstance(fail_test, bool):
+            raise AnsibleFilterError("json_patch: 'fail_test' argument is not a bool")
```
[the remark this answers is not in the document]
Comment (felixfontein) [reply plugins/filter/json_patch.py:142]: I don't think it's OK if the user passes in a string `"false"`, for example. The plugin should reject that, instead of treating it as `true`.

File: plugins/filter/json_patch.py [hunk 32]
```diff
@@ -0,0 +1,195 @@
...
+
+    def check_patch_arguments(self, filter_name: str, args: dict):
+
+        if "op" not in args or not isinstance(args["op"], str):
+            raise AnsibleFilterError(f"{filter_name}: 'op' argument is not a string")
+
+        if args["op"] not in OPERATIONS_AVAILABLE:
+            raise AnsibleFilterError(
+                f"{filter_name}: unsupported 'op' argument: {args['op']}"
+            )
```
[the remark this answers is not in the document]
Comment (felixfontein) [reply plugins/filter/json_patch.py:58]: Actually my comment was that the filter should do the validation :) The operations are specified in an RFC, so they won't just randomly change with a new library release.

File: plugins/filter/json_patch.yml (hunk 30 above)
[the remark this answers is not in the document]
Comment (numo68) [reply plugins/filter/json_patch.yml:12]: Searching in the code both are used approximately the same :) You guys decide (and perhaps the documentation should state that), I vote for `none` here.

File: plugins/filter/json_patch.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (numo68) [reply plugins/filter/json_patch.py:142]: I was thinking of it as well, duck typing versus catching the user doing things not intended. Not sure what I like better.

File: tests/unit/plugins/filter/test_json_patch.py [hunk 33]
```diff
@@ -0,0 +1,313 @@
...
+        with unittest.mock.patch(
+            "ansible_collections.community.general.plugins.filter.json_patch.HAS_LIB",
+            False,
+        ):
+            with self.assertRaises(AnsibleFilterError) as context:
+                self.json_patch_recipe({}, [])
+            self.assertEqual(
+                str(context.exception),
+                "You need to install 'jsonpatch' package prior to running 'json_patch_recipe' filter",
+            )
```
[the remark this answers is not in the document]
Comment (numo68) [reply tests/unit/plugins/filter/test_json_patch.py:219]: This comes from my day job - any line written should have line and branch coverage in the unit test, including error handling if possible. Things like copypaste errors are extremely common here. If I implement it, I test it.

File: plugins/filter/json_patch.py (hunk 32 above)
[the remark this answers is not in the document]
Comment (numo68) [reply plugins/filter/json_patch.py:58]: Yeah, Felix wanted me to test what is documented, and the set of operations is. So again, you guys decide. The underlying library outputs pretty good error messages so I don't see the absolute need here either.

File: tests/unit/plugins/filter/test_json_patch.py
[the remark this answers is not in the document]
Comment (numo68) [reply tests/unit/plugins/filter/test_json_patch.py:1]: On a "philosophical" note, the usual software development approach is that the unit tests are the ones that take care of thoroughly testing the functionality including the error handling. The integration ones test whether the components play together as intended. Again, I am not sure what the Ansible way is and I admit that the integration test has the advantage of possibly serving as an example of usage. But generally, if I am writing a Python code, I am using a unit testing framework that taps the interfaces being implemented directly.

So to me the "right" way is to have thorough unit tests first. I originally had that and did not see the necessity of an integration one at all, but seems that it is usual to have one. You are right that there is a redundancy now.

File: plugins/filter/json_patch.yml (hunk 30 above)
[the remark this answers is not in the document]
Comment (felixfontein) [reply plugins/filter/json_patch.yml:12]: Right now both `null` and `none` are not used at all in filter documentation. Since filters are Jinja2 we should use `none` since that's Jinja2, as opposed to `null`. `ansible localhost -m debug -a msg='{{ none | string }}'` works, `ansible localhost -m debug -a msg='{{ null | string }}'` does not.

[the remark this answers is not in the document]
Comment (felixfontein) [reply plugins/filter/json_patch.yml:12]: (While `None` also works, the test is called `none`: https://jinja.palletsprojects.com/en/stable/templates/#jinja-tests.none. Also see https://jinja.palletsprojects.com/en/stable/templates/#literals, in particular the 'Note' before the next section.)

File: plugins/filter/json_patch.py [hunk 34]
```diff
@@ -0,0 +1,195 @@
...
+        target = self.check_json_object("json_diff", "target", target)
+
+        try:
+            result = list(jsonpatch.make_patch(inp, target))
+        except Exception as e:
+            raise AnsibleFilterError(f"JSON diff failed: {e}") from e
+
+        return result
+
+    def filters(self) -> dict[str, Callable[[object, str, str, object], object]]:
```
[the remark this answers is not in the document]
Comment (numo68) [reply plugins/filter/json_patch.py:185]: Yep, this was wrong, changing to Callable[..., Any], probably no need to be extra strict here. I am unfortunately not an expert here and the type-hints are I think still kind of a moving target as the Python versions evolve.

File: tests/unit/plugins/filter/test_json_patch.py
[the remark this answers is not in the document]
Comment (numo68) [reply tests/unit/plugins/filter/test_json_patch.py:1]: Okay, leaving both for now, even if redundant. I don't want to remove/simplify the unit ones, as that's where I normally go when addressing bugs including error handling; reproduce in an unit-test first, then fix, and the integration ones are good for CI.

File: plugins/filter/json_patch.yml (hunk 30 above)
[the remark this answers is not in the document]
Comment (russoz) [reply plugins/filter/json_patch.yml:12]: Okie, I stand corrected.

File: plugins/filter/json_patch.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (russoz) [reply plugins/filter/json_patch.py:142]: Yeah, it's a conundrum.
 I don't have a strong argument either way so I will take it as it is.

File: tests/unit/plugins/filter/test_json_patch.py
[the remark this answers is not in the document]
Comment (russoz) [reply tests/unit/plugins/filter/test_json_patch.py:1]: Makes sense to me. Thanks!

File: tests/unit/plugins/filter/test_json_patch.py (hunk 33 above)
[the remark this answers is not in the document]
Comment (russoz) [reply tests/unit/plugins/filter/test_json_patch.py:219]: Well, still feels like overkill to me, but I'm not gonna press the issue. Thanks for you clarification.

File: tests/unit/requirements.txt [hunk 35]
```diff
@@ -56,4 +56,7 @@ python-nomad < 2.0.0 ; python_version <= '3.6'
 python-nomad >= 2.0.0 ; python_version >= '3.7'
 
 # requirement for jenkins_build, jenkins_node, jenkins_plugin modules
-python-jenkins >= 0.4.12
\ No newline at end of file
+python-jenkins >= 0.4.12
+
+# requirement for json_patch, json_patch_recipe and json_patch plugins
+jsonpatch
```
[the remark this answers is not in the document]
Comment (russoz) [reply tests/unit/requirements.txt:62]: This comment became redundant. :-)

**Review by russoz — APPROVED** [review]
LGTM

File: plugins/filter/json_patch.py [hunk 36]
```diff
@@ -0,0 +1,195 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) Stanislav Meduna (@numo68)
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from __future__ import absolute_import, annotations, division, print_function
```
Comment (felixfontein) [inline plugins/filter/json_patch.py:6]: I think you only need `annotations`, the rest is only needed for Python 2:
```suggestion
from __future__ import annotations
```

**Review by felixfontein — APPROVED** [review]
Looks good to me. If nobody objects, I'll merge this in a few days.
